google-cloud-firestore/src/main/java/com/google/cloud/firestore/GenericQuerySnapshot.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.firestore;
+
+import com.google.cloud.Timestamp;
+import com.google.cloud.firestore.encoding.CustomClassMapper;
+import com.google.common.collect.ImmutableList;
+import java.util.*;
+import javax.annotation.Nonnull;
+
+/**
+ * Abstract. A GenericQuerySnapshot represents the results of a query that returns documents. It can
+ * contain zero or more DocumentSnapshot objects.
+ */
+public abstract class GenericQuerySnapshot<QueryT> implements Iterable<QueryDocumentSnapshot> {
+  protected final QueryT query;
+  protected final Timestamp readTime;
+
+  volatile List<DocumentChange> documentChanges;
+  volatile List<QueryDocumentSnapshot> documents;
+
+  volatile DocumentSet documentSet;
+
+  // Elevated access level for mocking.
+  protected GenericQuerySnapshot(
+      QueryT query,
+      Timestamp readTime,
+      final List<QueryDocumentSnapshot> documents,
+      final DocumentSet documentSet,
```

Comment:
Also pre-existing logic, so I'm not certain of the reasoning, but there are several instances of lazy-loading and memoization in this code. My suspicion is that it's related to performance. I don't have strong feelings on this either way. What are your thoughts with that context?

Reply:
I changed this.

I don't think there was an issue before. The value of `documentSet` was tested for equality because it's value was converted by the `getDocuments()` method.

However, I think the new approach is equally fine. The benefits of lazily converting `documentSet` to a list of document snapshots are rarely if ever seen.